src/osd/PrimaryLogPG.cc

```diff
@@ -1641,18 +1641,36 @@ void PrimaryLogPG::calc_trim_to()
 	cct->_conf->osd_pg_log_trim_max >= cct->_conf->osd_pg_log_trim_min) {
       return;
     }
-    list<pg_log_entry_t>::const_iterator it = pg_log.get_log().log.begin();
+    auto it = pg_log.get_log().log.begin(); // oldest log entry
+    auto rit = pg_log.get_log().log.rbegin();
+    eversion_t candidate;
     eversion_t new_trim_to;
-    for (uint64_t i = 0; i < num_to_trim; ++i) {
-      new_trim_to = it->version;
-      ++it;
-      if (new_trim_to >= limit) {
-	new_trim_to = limit;
-        dout(10) << "calc_trim_to trimming to limit: " << limit << dendl;
-	break;
+    size_t i = 0;
+    while(it != pg_log.get_log().log.end()) {
+      i++;
+      if (i > target) {
+        new_trim_to = rit->version;
+        break;
+      }
+      if (i >= num_to_trim) {
+        candidate = it->version; // maybe
       }
+      it++;
+      rit++;
+    }
+    if (it == pg_log.get_log().log.end()) {
+      dout(10) << __func__ << " not enough log left to trim" << dendl;
+      return; // nothing to trim
+    }
+    if (new_trim_to > candidate) {
+      new_trim_to = candidate;
+    }
+    if (new_trim_to >= limit) {
+      new_trim_to = limit;
+      dout(10) << __func__ << " trimming to limit: " << limit << dendl;
     }
-    dout(10) << "calc_trim_to " << pg_trim_to << " -> " << new_trim_to << dendl;
+    dout(10) << __func__ << " " << pg_trim_to << " -> "
+             << new_trim_to << dendl;
```

Comment:
can we simplify this loop like
```c++
auto trim_to = pg_log.get_log().log.cbegin();
eversion_t trim_to_ver;
for (unsigned n_to_trim = 0, n_to_reserve = pg_log.get_log().log.size();
     n_to_trim < osd_pg_log_trim_max && n_to_reserve >= target;
     ++n_to_trim, --n_to_reserve) {
  if (trim_to == pg_log.get_log().log.cend()) {
    break;
  }
  trim_to_ver = trim_to->version;
  if (trim_to_ver >= limit) {
    break;
  }
  ++trim_to;
}
```

so we can stop the loop earlier once the `trim_to_ver` reaches the `limit`.

Reply:
also, i think we can update `pg_log.get_log().approx_size()` to take the advantage of O(1) implementation of `list::size()`. as we've moved to C++17, and GCC-7 is required.

if `approx_size()` is *accurate*, probably we don't need to fuss over this anymore.

Reply:
@tchaikov This change needs to be backported to mimc

Reply:
the same also applies to mimic.

Reply:
in master: 

https://github.com/ceph/ceph/blob/eac80144c82bcbe8fc3324b9e02672b84347fc54/src/CMakeLists.txt#L137-L146

in mimic:

https://github.com/ceph/ceph/blob/d46e68458241f9190b8b7e6af009d7e745134111/src/CMakeLists.txt#L184-L193